### PR TITLE
Sparse lanczos

### PR DESCRIPTION
**Pull request by mganahl**
implementation of sparse symmetric lanczos methods for numpy and pytorch backends

### CODE REVIEW

**Comment by mganahl** [comment]
Hey @Thenerdstation have a look and let me know what you think!

**Comment by mganahl** [comment]
I'll close the other PR since its obsolete

File: tensornetwork/backends/numpy/numpy_backend.py [hunk 1]
```diff
@@ -135,14 +136,136 @@ def randn(self,
...
+    if ((dtype is self.np.dtype(self.np.complex128)) or
+        (dtype is self.np.dtype(self.np.complex64))):
       return self.np.random.randn(*shape).astype(
           dtype) + 1j * self.np.random.randn(*shape).astype(dtype)
     return self.np.random.randn(*shape).astype(dtype)
 
   def conj(self, tensor: Tensor) -> Tensor:
     return self.np.conj(tensor)
+
+  def eigsh_lanczos(
```
Comment (Thenerdstation) [inline tensornetwork/backends/numpy/numpy_backend.py]: Be sure to add this to `BaseBackend` too

Comment (mganahl) [reply tensornetwork/backends/numpy/numpy_backend.py]: added eigsh_lanczos to all other backends.

File: tensornetwork/backends/base_backend.py [hunk 2]
```diff
@@ -285,3 +284,43 @@ def conj(self, tensor: Tensor) -> Tensor:
...
+    of `A`. 
+    Args:
+      A: A (sparse) implementation of a linear operator. 
+      initial_state: An initial vector for the Lanczos algorithm. If `None`,
+        a random initial `Tensor` is created using the `torch.randn` method
+      ncv: The number of iterations (number of krylov vectors).
+      numeig: The nummber of eigenvector-eigenvalue pairs to be computed.
+        If `numeig > 1`, `reorthogonalize` has to be `True`.
+      tol: The desired precision of the eigenvalus. Currently we use 
+        `torch.norm(eigvalsnew[0:numeig] - eigvalsold[0:numeig]) < tol`
```
Comment (Thenerdstation) [inline tensornetwork/backends/base_backend.py]: Don't mention `torch` in the base backend.

File: tensornetwork/backends/numpy/numpy_backend.py [hunk 3]
```diff
@@ -135,14 +136,135 @@ def randn(self,
...
+      ndiag: The tridiagonal Operator is diagonalized every `ndiag` iterations 
+        to check convergence.
+      reorthogonalize: If `True`, Krylov vectors are kept orthogonal by 
+        explicit orthogonalization (more costly than `reorthogonalize=False`)
+    Returns:
+      (eigvals, eigvecs)
+       eigvals: A list of `numeig` lowest eigenvalues
+       eigvecs: A list of `numeig` lowest eigenvectors
+    """
+    #TODO: make this work for tensorflow in graph mode
```
Comment (Thenerdstation) [inline tensornetwork/backends/numpy/numpy_backend.py]: Remove todo for numpy backend.

File: tensornetwork/backends/shell/shell_backend.py [hunk 4]
```diff
@@ -210,3 +209,16 @@ def randn(self,
...
+      self,
+      A: Callable,
+      initial_state: Optional[Tensor] = None,
+      ncv: Optional[int] = 200,
+      numeig: Optional[int] = 1,
+      tol: Optional[float] = 1E-8,
+      delta: Optional[float] = 1E-8,
+      ndiag: Optional[int] = 20,
+      reorthogonalize: Optional[bool] = False) -> Tuple[List, List]:
+    raise NotImplementedError(
```
Comment (Thenerdstation) [inline tensornetwork/backends/shell/shell_backend.py]: It should (?) be possible to add this to the shell backend. It's just shape calculations.

Comment (mganahl) [reply tensornetwork/backends/shell/shell_backend.py]: added

File: tensornetwork/backends/base_backend.py (hunk 2 above)
Comment (mganahl) [reply tensornetwork/backends/base_backend.py]: removed

File: tensornetwork/backends/numpy/numpy_backend.py (hunk 3 above)
Comment (mganahl) [reply tensornetwork/backends/numpy/numpy_backend.py]: removed

**Pull request merged**